Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+        m_stringBuilder.append("unsigned");
+    else
+        m_stringBuilder.append(type.name());
+}
+
+void FunctionDefinitionWriter::visit(AST::ParameterizedType& type)
+{
+    switch (type.base()) {
+    case AST::ParameterizedType::Base::Vec2:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 2>");
+        break;
+    case AST::ParameterizedType::Base::Vec3:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 3>");
+        break;
+    case AST::ParameterizedType::Base::Vec4:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 4>");
+        break;
+    case AST::ParameterizedType::Base::Mat2x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat2x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat2x4:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x4:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x4:
+        // TODO
+        break;
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Parameter& parameter)
+{
+    visit(parameter.type());
+    m_stringBuilder.append(" ", parameter.name());
+    for (auto& attribute : parameter.attributes()) {
+        m_stringBuilder.append(" ");
+        checkErrorAndVisit(attribute);
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Expression& expression)
+{
+    AST::Visitor::visit(expression);
+}
+
+void FunctionDefinitionWriter::visit(AST::CallableExpression& call)
+{
+    bool first = true;
+    if (call.target().kind() == AST::Node::Kind::ArrayType) {
+        m_stringBuilder.append("{\n");
+        {
+            IndentationScope scope(m_indent);
+            for (auto& argument : call.arguments()) {
+                m_stringBuilder.append(m_indent);
+                visit(argument);
+                m_stringBuilder.append(",\n");
+                first = false;
+            }
+        }
+        m_stringBuilder.append(m_indent, "}");
+    } else {
+        visit(call.target());
+        m_stringBuilder.append("(");
+        for (auto& argument : call.arguments()) {
+            if (!first)
+                m_stringBuilder.append(", ");
+            visit(argument);
+            first = false;
+        }
+        m_stringBuilder.append(")");
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::UnaryExpression& unary)
+{
+    switch (unary.operation()) {
+    case AST::UnaryOperation::Negate:
+        m_stringBuilder.append("-");
+    }
+    visit(unary.expression());
```

Comment:
I think we should spray parentheses everywhere, to make sure order of operations is preserved.